olive/data/component/pre_process_data.py

```diff
@@ -122,7 +122,119 @@ def _tokenizer_and_align_labels(examples):
         tokenized_inputs["label"] = new_labels
         return tokenized_inputs
 
-    tokenized_datasets = _huggingface_pre_precess_helper(
+    tokenized_datasets = _huggingface_pre_process_helper(
         _dataset, model_name, input_cols, label_cols, _tokenizer_and_align_labels, **kwargs
     )
     return BaseDataset(tokenized_datasets, label_cols=["label"], max_samples=max_samples)
+
+
+@Registry.register_pre_process()
+def text_generation_huggingface_pre_process(
+    _dataset, model_name, input_cols, seqlen, stride=None, max_samples=None, **kwargs
+):
+    """Pre-process data.
+
+    Args:
+        _dataset (object): Data to be pre-processed.
+        model_name (str): Name of the huggingface model.
+        input_cols (list): List of input columns.
+        seqlen (int): Length of the sequence.
+        stride (int): Stride to use when splitting the sequence. I
+            If None, the sequence is split into non-overlapping sequences. No context is used.
+            When stride is not None, we use a sliding window approach to split the sequence. The stride is
+            also used as context length.
+        max_samples (int): Maximum number of samples to use.
+        **kwargs: Additional arguments.
+            joiner (str): Delimiter to use when joining the rows of the input columns.
+            random_seed (int): Random seed to use. If not None, we use the random seed to choose the starting
+                point of the sequence.
+
+    Returns:
+        object: Pre-processed data.
+    """
+    from random import Random
+
+    from datasets import Dataset as HFDataset
+    from transformers import AutoTokenizer
+
+    # get tokenizer
+    tokenizer = AutoTokenizer.from_pretrained(model_name)
+
+    # gather text from all input columns
+    text_list = []
+    for input_col in input_cols:
+        text_list += _dataset[input_col]
+    # delimiter between the text sequences
+    joiner = kwargs.get("joiner", " ")
+    text = joiner.join(text_list)
+
+    # in order to make processing faster we will only tokenize as much as needed
+    # assumes that num words > num tokens
+    split_text = text.split(" ")
+    num_text = len(split_text)
+
+    tokenized_inputs = {
+        "input_ids": [],
+        "target_ids": [],
+    }
+
+    random_seed = kwargs.get("random_seed")
+    if random_seed is None:
+        # no randomization, just use contiguous blocks of tokens
+        max_text = max_samples * seqlen if max_samples is not None else num_text
+        max_text = min(max_text, num_text)
+        encodings = tokenizer(" ".join(split_text[:max_text]), return_tensors="pt")
```

Comment:
It may hurt PPL value if we drops some inputs, especially for some popular model(gpt2), user may say why the PPL value is different with the public report.

Or maybe you consider this somewhere?

Reply:
Do you mean the `max_samples` dropping inputs? If users want to use the whole dataset, they can set `max_samples` to `None`. And it will use the entire text.